### DISCUSSION TITLE
Support for building and distributing Typescript types for Meteor Packages

### DISCUSSION THREAD

**Discussion by perbergland** [meteor/meteor#11735] [Feature Requests]
Today the only way to get types for Meteor packages is to write them "on the side". Some types are on [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped), e.g. the ones for Meteor itself, whereas others are in gists or just shared peer to peer on forums or slack.

This is a pain point for developers that want to use Typescript for meteor projects.

npm packages support typescript by including .d.ts files in the distribution that get unpacked into node_modules and then picked up by the typescript compiler.

Several changes would need to be made to allow for publishing types as part of Meteor Packages:

* Build process support and some way to include the produced .d.ts file(s) in the meteor published bundle. Meteor packages compared to apps have no package.json support today so the easiest way is probably to augment package.js and isobuild in some way.
* The produced files then need to be made available to the Typescript compiler. Since Meteor compared to npm doesn’t unpack packages locally in the project we would need to add a new paradigm to put the .d.ts in the project directory in a well established directory similar to node_modules

**Comment by perbergland** [comment]
I have never authored a single npm package, let alone with typescript, so looking forward to someone writing up some best practices here that we can be inspired by.

**Comment by radekmie** [comment]
As one of the most active `@types/meteor` maintainers, as well as the `universe:i18n` maintainer ([repo](https://github.com/vazco/meteor-universe-i18n/), [types](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/meteor-universe-i18n)), I have to say it'd be nice to provide typings for Atmosphere packages directly in the package. This includes core packages, like `meteor/mongo`, of course.

There was [this](https://forums.meteor.com/t/attempt-at-getting-d-ts-files-in-meteor-packages-to-be-usefull/56368/1) attempt at making it possible, but I'm not sure whether that's the correct approach - changing `node_modules` is not a good idea in general. I do, however, see an option to leverage the [`typeRoots`](https://www.typescriptlang.org/tsconfig#typeRoots) option.

**Comment by JanMP** [comment]
I switched back to coffescript (after almost a year of TS). There's a relatively new VS Code plugin that offers the same intellisense support that vanilla JS gets. I am just starting to add type-information and comments in jsdoc comments wich is really nice because I now got my uncluttered coffescript with all the additional info tucked away in block comments (that can be folded into a neat little ###*). It would be really great to have that working across meteor packages.
